Libm/man3/math.3
.\" Copyright (c) 1991 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     from: @(#)acos.3	5.1 (Berkeley) 5/2/91
.\"	$Id: math.3,v 1.2 2003/08/17 20:36:47 scp Exp $
.\"
.Dd March 20, 2007
.Dt MATH 3
.Os
.Sh NAME
.Nm math
.Nd mathematical library functions
.Sh SYNOPSIS
.Fd #include <math.h>
.Sh DESCRIPTION
The header file math.h provides function prototypes and macros for working with
C99 floating point values.
.Pp
Each math.h function is provided in three variants: single, double and extended precision.
The single and double precision variants operate on IEEE-754 single and double precision
values, which correspond to the C types
.Ft float
and
.Ft double ,
respectively.
.Pp
On Intel Macs, the C type
.Ft long double
corresponds to 80-bit IEEE-754 double extended precision.  On PowerPC Macs, the C type
.Ft long double
corresponds by default to ordinary double precision, but a 128-bit non-IEEE-754 long double
type is also available via the compiler flag
.Ft -mlong-double-128 
and linker flag
.Ft -lmx .
.Pp
Details of the floating point formats can be found via "man float".
.Pp
Users who need to repeatedly perform the same calculation on a large set of data will
probably find that the vector math library (composed of vMathLib and vForce) yields
better performance for their needs than sequential calls to the libm.
.Pp
Users who need to perform mathematical operations on complex floating-point numbers should
consult the man pages for the complex portion of the math library, via "man complex".
.Sh LIST OF FUNCTIONS
Each of the functions that use floating-point values are provided in single, double, 
and extended precision; the double precision prototypes are listed here.  The man
pages for the individual functions provide more details on their use, special cases, and
prototypes for their single and extended precision versions.
.Pp
.Ft int
.Fn fpclassify "double"
.br
.Ft int
.Fn isfinite "double"
.br
.Ft int
.Fn isinf "double"
.br
.Ft int
.Fn isnan "double"
.br
.Ft int
.Fn isnormal "double"
.br
.Ft int
.Fn signbit "double"
.Pp
These function-like macros are used to classify a single floating-point argument.
.Pp
.Ft double
.Fn copysign "double, double"
.br
.Ft double
.Fn nextafter "double, double"
.Pp
.Fn copysign "x, y"
returns the value equal in magnitude to 
.Fa x
with the sign of
.Fa y .
.Fn nextafter "x, y"
returns the next floating-point number after
.Fa x
in the direction of
.Fa y .
Both are correctly-rounded.
.Pp
.Ft double
.Fn nan "const char *tag"
.Pp
The
.Fn nan
function returns a quiet NaN, without raising the invalid flag.
.Pp
.Ft double
.Fn ceil "double"
.br
.Ft double
.Fn floor "double"
.br
.Ft double
.Fn nearbyint "double"
.br
.Ft double
.Fn rint "double"
.br
.Ft double
.Fn round "double"
.br
.Ft long int
.Fn lrint "double"
.br
.Ft long int
.Fn lround "double"
.br
.Ft long long int
.Fn llrint "double"
.br
.Ft long long int
.Fn llround "double"
.br
.Ft double
.Fn trunc "double"
.Pp
These functions provide various means to round floating-point values to integral values.  They are correctly rounded.
.Pp
.Ft double
.Fn fmod "double, double"
.br
.Ft double
.Fn remainder "double, double"
.br
.Ft double
.Fn remquo "double x, double y, int *"
.Pp
These return a remainder of the division of x by y with an integral quotient.
.Fn remquo
additionally provides access to a few lower bits of the quotient.  They are correctly rounded.
.Pp
.Ft double
.Fn fdim "double, double"
.br
.Ft double
.Fn fmax "double, double"
.br
.Ft double
.Fn fmin "double, double"
.Pp
.Fn fmax "x, y"
and
.Fn fmin "x, y"
return the maximum and minimum of
.Fa x
and
.Fa y ,
respectively.
.Fn fdim "x, y"
returns the positive difference of
.Fa x 
and
.Fa y .  All are correctly rounded.
.Pp
.Ft double
.Fn fma "double x, double y, double z"
.Pp
.Fn fma "x, y, z"
computes the value (x*y) + z as though without intermediate rounding.  It is correctly rounded.
.Pp
.Ft double
.Fn fabs "double"
.br
.Ft double
.Fn sqrt "double"
.br
.Ft double
.Fn cbrt "double"
.br
.Ft double
.Fn hypot "double, double"
.Pp
.Fn fabs "x",
.Fn sqrt "x",
and
.Fn cbrt "x"
return the absolute value, square root, and cube root of
.Fa x ,
respectively.
.Fn hypot "x, y"
returns sqrt(x*x + y*y).
.Fn fabs
and
.Fn sqrt
are correctly rounded.
.Pp
.Ft double
.Fn exp "double"
.br
.Ft double
.Fn exp2 "double"
.br
.Ft double
.Fn expm1 "double"
.Pp
.Fn exp "x" ,
.Fn exp2 "x" ,
and
.Fn expm1 "x"
return e**x, 2**x, and e**x - 1, respectively.
.Pp
.Ft double
.Fn log "double"
.br
.Ft double
.Fn log2 "double"
.br
.Ft double
.Fn log10 "double"
.br
.Ft double
.Fn log1p "double"
.Pp
.Fn log "x" ,
.Fn log2 "x" ,
and
.Fn log10 "x"
return the natural, base-2, and base-10 logarithms of
.Fa x ,
respectively.
.Fn log1p "x"
returns the natural log of 1+x.
.Pp
.Ft double
.Fn logb "double"
.br
.Ft int
.Fn ilogb "double"
.Pp
.Fn logb "x"
and
.Fn ilogb "x"
return the exponent of
.Fa x .
.Pp
.Ft double
.Fn modf "double, double *"
.br
.Ft double
.Fn frexp "double, int *"
.Pp
.Fn modf "x, &y"
returns the fractional part of
.Fa x
and stores the integral part in
.Fa y .
.Fn frexp "x, &n"
returns the mantissa of
.Fa x
and stores the exponent in
.Fa n .  They are correctly rounded.
.Pp
.Ft double
.Fn ldexp "double, int"
.br
.Ft double
.Fn scalbn "double, int"
.br
.Ft double
.Fn scalbln "double, long int"
.Pp
.Fn ldexp "x, n" ,
.Fn scalbn "x, n" ,
and
.Fn scalbln "x, n"
return x*2**n.  They are correctly rounded.
.Pp
.Ft double
.Fn pow "double, double"
.Pp
.Fn pow "x,y"
returns x raised to the power y.
.Pp
.Ft double
.Fn cos "double"
.br
.Ft double
.Fn sin "double"
.br
.Ft double
.Fn tan "double"
.Pp
.Fn cos "x" ,
.Fn sin "x" ,
and
.Fn tan "x"
return the cosine, sine and tangent of
.Fa x ,
respectively.
.Pp
.Ft double
.Fn cosh "double"
.br
.Ft double
.Fn sinh "double"
.br
.Ft double
.Fn tanh "double"
.Pp
.Fn cosh "x" ,
.Fn sinh "x" ,
and
.Fn tanh "x"
return the hyperbolic cosine, hyperbolic sine and hyperbolic tangent of
.Fa x ,
respectively.
.Pp
.Ft double
.Fn acos "double"
.br
.Ft double
.Fn asin "double"
.br
.Ft double
.Fn atan "double"
.br
.Ft double
.Fn atan2 "double, double"
.Pp
.Fn acos "x" ,
.Fn asin "x" ,
and
.Fn atan "x"
return the inverse cosine, inverse sine and inverse tangent of
.Fa x ,
respectively.
.Fn atan2 "y, x"
returns the inverse tangent of y/x, with sign chosen according to the quadrant of (x,y).
.Pp
.Ft double
.Fn acosh "double"
.br
.Ft double
.Fn asinh "double"
.br
.Ft double
.Fn atanh "double"
.Pp
.Fn acosh "x" ,
.Fn asinh "x" ,
and
.Fn atanh "x"
return the inverse hyperbolic cosine, inverse hyperbolic sine and inverse hyperbolic tangent of
.Fa x ,
respectively.
.Pp
.Ft double
.Fn tgamma "double"
.br
.Ft double
.Fn lgamma "double"
.Pp
.Fn tgamma "x"
and
.Fn lgamma "x"
return the values of the gamma function and its logarithm evalutated at
.Fa x ,
respectively.
.Pp
.Ft double
.Fn j0 "double"
.br
.Ft double
.Fn j1 "double"
.br
.Ft double
.Fn jn "double"
.br
.Ft double
.Fn y0 "double"
.br
.Ft double
.Fn y1 "double"
.br
.Ft double
.Fn yn "double"
.Pp
.Fn j0 "x" ,
.Fn j1 "x" ,
and
.Fn jn "x"
return the values of the zeroth, first, and nth Bessel function of the first kind evaluated at
.Fa x ,
respectively.
.Fn y0 "x" ,
.Fn y1 "x" ,
and
.Fn yn "x"
return the values of the zeroth, first, and nth Bessel function of the second kind evaluated at
.Fa x ,
respectively.
.Pp
.Ft double
.Fn erf "double"
.br
.Ft double
.Fn erfc "double"
.Pp
.Fn erf "x"
and
.Fn erfc "x"
return the values of the error function and the complementary error function evaluated at
.Fa x ,
respectively.
.Sh MATHEMATICAL CONSTANTS
In addition to the functions listed above, math.h defines a number of useful constants,
listed below.  All are defined as C99 floating-point constants.
.nf
.ta \w'M_2_SQRTPI'u+6
.sp 1
CONSTANT	VALUE
M_E	base of natural logarithm, e
M_LOG2E	log2(e)
M_LOG10E	log10(e)
M_LN2	ln(2)
M_LN10	ln(10)
M_PI	pi
M_PI_2	pi / 2
M_PI_4	pi / 4
M_1_PI	1 / pi
M_2_PI	2 / pi
M_2_SQRTPI	2 / sqrt(pi)
M_SQRT2	sqrt(2)
M_SQRT1_2	sqrt(1/2)
.ta
.fi
.Sh IEEE STANDARD 754 FLOATING\-POINT ARITHMETIC
The libm functions declared in math.h provide mathematical library functions in
single-, double-, and extended-precision IEEE-754 floating-point formats on Intel macs,
and in single- and double-precision IEEE-754 floating-point formats on PowerPC macs.
.Sh SEE ALSO
.Xr float 3 ,
.Xr complex 3
.Sh STANDARDS
The <math.h> functions conform to the ISO/IEC 9899:1999(E) standard.

X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


ruby/ruby/math.c
/**********************************************************************

  math.c -

  $Author: shyouhei $
  $Date: 2007-02-13 08:01:19 +0900 (Tue, 13 Feb 2007) $
  created at: Tue Jan 25 14:12:56 JST 1994

  Copyright (C) 1993-2003 Yukihiro Matsumoto

**********************************************************************/

#include "ruby.h"
#include <math.h>
#include <errno.h>

VALUE rb_mMath;

#define Need_Float(x) (x) = rb_Float(x)
#define Need_Float2(x,y) do {\
    Need_Float(x);\
    Need_Float(y);\
} while (0)

static void
domain_check(x, msg)
    double x;
    char *msg;
{
    while(1) {
	if (errno) {
	    rb_sys_fail(msg);
	}
	if (isnan(x)) {
#if defined(EDOM)
	    errno = EDOM;
#elif define(ERANGE)
	    errno = ERANGE;
#endif
	    continue;
	}
	break;
    }
}


/*
 *  call-seq:
 *     Math.atan2(y, x)  => float
 *  
 *  Computes the arc tangent given <i>y</i> and <i>x</i>. Returns
 *  -PI..PI.
 *     
 */

static VALUE
math_atan2(obj, y, x)
    VALUE obj, x, y;
{
    Need_Float2(y, x);
    return rb_float_new(atan2(RFLOAT(y)->value, RFLOAT(x)->value));
}


/*
 *  call-seq:
 *     Math.cos(x)    => float
 *  
 *  Computes the cosine of <i>x</i> (expressed in radians). Returns
 *  -1..1.
 */

static VALUE
math_cos(obj, x)
    VALUE obj, x;
{
    Need_Float(x);
    return rb_float_new(cos(RFLOAT(x)->value));
}

/*
 *  call-seq:
 *     Math.sin(x)    => float
 *  
 *  Computes the sine of <i>x</i> (expressed in radians). Returns
 *  -1..1.
 */

static VALUE
math_sin(obj, x)
    VALUE obj, x;
{
    Need_Float(x);

    return rb_float_new(sin(RFLOAT(x)->value));
}


/*
 *  call-seq:
 *     Math.tan(x)    => float
 *  
 *  Returns the tangent of <i>x</i> (expressed in radians).
 */

static VALUE
math_tan(obj, x)
    VALUE obj, x;
{
    Need_Float(x);

    return rb_float_new(tan(RFLOAT(x)->value));
}

/*
 *  call-seq:
 *     Math.acos(x)    => float
 *  
 *  Computes the arc cosine of <i>x</i>. Returns 0..PI.
 */

static VALUE
math_acos(obj, x)
    VALUE obj, x;
{
    double d;

    Need_Float(x);
    errno = 0;
    d = acos(RFLOAT(x)->value);
    domain_check(d, "acos");
    return rb_float_new(d);
}

/*
 *  call-seq:
 *     Math.asin(x)    => float
 *  
 *  Computes the arc sine of <i>x</i>. Returns 0..PI.
 */

static VALUE
math_asin(obj, x)
    VALUE obj, x;
{
    double d;

    Need_Float(x);
    errno = 0;
    d = asin(RFLOAT(x)->value);
    domain_check(d, "asin");
    return rb_float_new(d);
}

/*
 *  call-seq:
 *     Math.atan(x)    => float
 *  
 *  Computes the arc tangent of <i>x</i>. Returns -{PI/2} .. {PI/2}.
 */

static VALUE
math_atan(obj, x)
    VALUE obj, x;
{
    Need_Float(x);
    return rb_float_new(atan(RFLOAT(x)->value));
}

#ifndef HAVE_COSH
double
cosh(x)
    double x;
{
    return (exp(x) + exp(-x)) / 2;
}
#endif

/*
 *  call-seq:
 *     Math.cosh(x)    => float
 *  
 *  Computes the hyperbolic cosine of <i>x</i> (expressed in radians).
 */

static VALUE
math_cosh(obj, x)
    VALUE obj, x;
{
    Need_Float(x);
    
    return rb_float_new(cosh(RFLOAT(x)->value));
}

#ifndef HAVE_SINH
double
sinh(x)
    double x;
{
    return (exp(x) - exp(-x)) / 2;
}
#endif

/*
 *  call-seq:
 *     Math.sinh(x)    => float
 *  
 *  Computes the hyperbolic sine of <i>x</i> (expressed in
 *  radians).
 */

static VALUE
math_sinh(obj, x)
    VALUE obj, x;
{
    Need_Float(x);
    return rb_float_new(sinh(RFLOAT(x)->value));
}

#ifndef HAVE_TANH
double
tanh(x)
    double x;
{
    return sinh(x) / cosh(x);
}
#endif

/*
 *  call-seq:
 *     Math.tanh()    => float
 *  
 *  Computes the hyperbolic tangent of <i>x</i> (expressed in
 *  radians).
 */

static VALUE
math_tanh(obj, x)
    VALUE obj, x;
{
    Need_Float(x);
    return rb_float_new(tanh(RFLOAT(x)->value));
}

/*
 *  call-seq:
 *     Math.acosh(x)    => float
 *  
 *  Computes the inverse hyperbolic cosine of <i>x</i>.
 */

static VALUE
math_acosh(obj, x)
    VALUE obj, x;
{
    double d;

    Need_Float(x);
    errno = 0;
    d = acosh(RFLOAT(x)->value);
    domain_check(d, "acosh");
    return rb_float_new(d);
}

/*
 *  call-seq:
 *     Math.asinh(x)    => float
 *  
 *  Computes the inverse hyperbolic sine of <i>x</i>.
 */

static VALUE
math_asinh(obj, x)
    VALUE obj, x;
{
    Need_Float(x);
    return rb_float_new(asinh(RFLOAT(x)->value));
}

/*
 *  call-seq:
 *     Math.atanh(x)    => float
 *  
 *  Computes the inverse hyperbolic tangent of <i>x</i>.
 */

static VALUE
math_atanh(obj, x)
    VALUE obj, x;
{
    double d;

    Need_Float(x);
    errno = 0;
    d = atanh(RFLOAT(x)->value);
    domain_check(d, "atanh");
    return rb_float_new(d);
}

/*
 *  call-seq:
 *     Math.exp(x)    => float
 *  
 *  Returns e**x.
 */

static VALUE
math_exp(obj, x)
    VALUE obj, x;
{
    Need_Float(x);
    return rb_float_new(exp(RFLOAT(x)->value));
}

#if defined __CYGWIN__
# include <cygwin/version.h>
# if CYGWIN_VERSION_DLL_MAJOR < 1005
#  define nan(x) nan()
# endif
# define log(x) ((x) < 0.0 ? nan("") : log(x))
# define log10(x) ((x) < 0.0 ? nan("") : log10(x))
#endif

/*
 *  call-seq:
 *     Math.log(numeric)    => float
 *  
 *  Returns the natural logarithm of <i>numeric</i>.
 */

static VALUE
math_log(obj, x)
    VALUE obj, x;
{
    double d;

    Need_Float(x);
    errno = 0;
    d = log(RFLOAT(x)->value);
    domain_check(d, "log");
    return rb_float_new(d);
}

/*
 *  call-seq:
 *     Math.log10(numeric)    => float
 *  
 *  Returns the base 10 logarithm of <i>numeric</i>.
 */

static VALUE
math_log10(obj, x)
    VALUE obj, x;
{
    double d;

    Need_Float(x);
    errno = 0;
    d = log10(RFLOAT(x)->value);
    domain_check(d, "log10");
    return rb_float_new(d);
}

/*
 *  call-seq:
 *     Math.sqrt(numeric)    => float
 *  
 *  Returns the non-negative square root of <i>numeric</i>.
 */

static VALUE
math_sqrt(obj, x)
    VALUE obj, x;
{
    double d;

    Need_Float(x);
    errno = 0;
    d = sqrt(RFLOAT(x)->value);
    domain_check(d, "sqrt");
    return rb_float_new(d);
}

/*
 *  call-seq:
 *     Math.frexp(numeric)    => [ fraction, exponent ]
 *  
 *  Returns a two-element array containing the normalized fraction (a
 *  <code>Float</code>) and exponent (a <code>Fixnum</code>) of
 *  <i>numeric</i>.
 *     
 *     fraction, exponent = Math.frexp(1234)   #=> [0.6025390625, 11]
 *     fraction * 2**exponent                  #=> 1234.0
 */

static VALUE
math_frexp(obj, x)
    VALUE obj, x;
{
    double d;
    int exp;

    Need_Float(x);
    
    d = frexp(RFLOAT(x)->value, &exp);
    return rb_assoc_new(rb_float_new(d), INT2NUM(exp));
}

/*
 *  call-seq:
 *     Math.ldexp(flt, int) -> float
 *  
 *  Returns the value of <i>flt</i>*(2**<i>int</i>).
 *     
 *     fraction, exponent = Math.frexp(1234)
 *     Math.ldexp(fraction, exponent)   #=> 1234.0
 */

static VALUE
math_ldexp(obj, x, n)
    VALUE obj, x, n;
{
    Need_Float(x);
    return rb_float_new(ldexp(RFLOAT(x)->value, NUM2INT(n)));
}

/*
 *  call-seq:
 *     Math.hypot(x, y)    => float
 *  
 *  Returns sqrt(x**2 + y**2), the hypotenuse of a right-angled triangle
 *  with sides <i>x</i> and <i>y</i>.
 *     
 *     Math.hypot(3, 4)   #=> 5.0
 */

static VALUE
math_hypot(obj, x, y)
    VALUE obj, x, y;
{
    Need_Float2(x, y);
    return rb_float_new(hypot(RFLOAT(x)->value, RFLOAT(y)->value));
}

/*
 * call-seq:
 *    Math.erf(x)  => float
 *
 *  Calculates the error function of x.
 */

static VALUE
math_erf(obj, x)
    VALUE obj, x;
{
    Need_Float(x);
    return rb_float_new(erf(RFLOAT(x)->value));
}

/*
 * call-seq:
 *    Math.erfc(x)  => float
 *
 *  Calculates the complementary error function of x.
 */

static VALUE
math_erfc(obj, x)
    VALUE obj, x;
{
    Need_Float(x);
    return rb_float_new(erfc(RFLOAT(x)->value));
}

/*
 *  The <code>Math</code> module contains module functions for basic
 *  trigonometric and transcendental functions. See class
 *  <code>Float</code> for a list of constants that
 *  define Ruby's floating point accuracy.
 */     


void
Init_Math()
{
    rb_mMath = rb_define_module("Math");

#ifdef M_PI
    rb_define_const(rb_mMath, "PI", rb_float_new(M_PI));
#else
    rb_define_const(rb_mMath, "PI", rb_float_new(atan(1.0)*4.0));
#endif

#ifdef M_E
    rb_define_const(rb_mMath, "E", rb_float_new(M_E));
#else
    rb_define_const(rb_mMath, "E", rb_float_new(exp(1.0)));
#endif

    rb_define_module_function(rb_mMath, "atan2", math_atan2, 2);
    rb_define_module_function(rb_mMath, "cos", math_cos, 1);
    rb_define_module_function(rb_mMath, "sin", math_sin, 1);
    rb_define_module_function(rb_mMath, "tan", math_tan, 1);

    rb_define_module_function(rb_mMath, "acos", math_acos, 1);
    rb_define_module_function(rb_mMath, "asin", math_asin, 1);
    rb_define_module_function(rb_mMath, "atan", math_atan, 1);

    rb_define_module_function(rb_mMath, "cosh", math_cosh, 1);
    rb_define_module_function(rb_mMath, "sinh", math_sinh, 1);
    rb_define_module_function(rb_mMath, "tanh", math_tanh, 1);

    rb_define_module_function(rb_mMath, "acosh", math_acosh, 1);
    rb_define_module_function(rb_mMath, "asinh", math_asinh, 1);
    rb_define_module_function(rb_mMath, "atanh", math_atanh, 1);

    rb_define_module_function(rb_mMath, "exp", math_exp, 1);
    rb_define_module_function(rb_mMath, "log", math_log, 1);
    rb_define_module_function(rb_mMath, "log10", math_log10, 1);
    rb_define_module_function(rb_mMath, "sqrt", math_sqrt, 1);

    rb_define_module_function(rb_mMath, "frexp", math_frexp, 1);
    rb_define_module_function(rb_mMath, "ldexp", math_ldexp, 2);

    rb_define_module_function(rb_mMath, "hypot", math_hypot, 2);

    rb_define_module_function(rb_mMath, "erf",  math_erf,  1);
    rb_define_module_function(rb_mMath, "erfc", math_erfc, 1);
}
